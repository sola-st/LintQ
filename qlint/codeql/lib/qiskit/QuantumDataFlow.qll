import semmle.python.dataflow.new.DataFlow
import semmle.python.dataflow.new.TaintTracking
import semmle.python.ApiGraphs
import qiskit.Circuit
import qiskit.QuantumOperator


/** Holds if the second operator may follow the first one. */
bindingset[start, file]
pragma[inline]
predicate mayFollow(QuantumOperator start, QuantumOperator end, File file, int qubitIndex) {
  // they are in the same file
  start.getLocation().getFile() = file and
  end.getLocation().getFile() = file and
  // they belong to the same circuit
  start.getQuantumCircuit() = end.getQuantumCircuit() and
  // they are connected via the control flow
  exists(
    QuantumCircuit circ, QubitUse startBitUse, QubitUse endBitUse, ControlFlowNode startNode,
    ControlFlowNode endNode, ControlFlowNode circNode
  |
    // they belong to the same circuit
    circ = start.getQuantumCircuit() and
    circ = end.getQuantumCircuit() and
    // they are connected to the control flow
    startNode = start.getNode() and
    endNode = end.getNode() and
    // they are applied in the right order: start >> end
    startNode.strictlyReaches(endNode) and
    // connect bit use and the two gates
    startBitUse.getAGate() = start and
    endBitUse.getAGate() = end and
    // and they act on the same position
    // bind the qubit index
    startBitUse.getAnIndex() = qubitIndex and
    endBitUse.getAnIndex() = qubitIndex and
    // exclude undefined bit indices
    qubitIndex >= 0 and
    (
      // they act on the same register
      // (if there is one explicitely instantiated)
      startBitUse.getARegister() = endBitUse.getARegister()
      or
      // or they act on the single quantum register of a circuit
      // experessed implicitely with e.g. QuantumCircuit(4)
      count(QuantumRegisterV2 reg | reg = circ.getAQuantumRegister() | reg) = 0 and
      circ.getNumberOfQubits() > 0
      or
      // or they act on the same quantum register but one uses the
      // integer only and the other uses the register object
      // this is unambiguous only with a single register
      // qc.h(0)
      // qc.x(qreg[0])
      count(QuantumRegisterV2 reg | reg = circ.getAQuantumRegister() | reg) = 1 and
      circ.getNumberOfQubits() > 0
    ) and
    // EXTRA PRECISION
    // they refer to the same circuit instance
    circNode = circ.getNode() and
    circNode.strictlyReaches(startNode) and
    circNode.strictlyReaches(endNode) and
    // we do not want a situation where the order is:
    // other >> initialization >> gate
    // because they would not refer to the same circuit anymore
    not startNode.strictlyReaches(circNode)
  )
}

/** Holds if the two operators refer to the same qubit. */
bindingset[opA, file]
pragma[inline]
predicate manipulateSameQubit(QuantumOperator opA, QuantumOperator opB, QuantumCircuit qc, int i, File file) {
  exists(
    QubitUse qbuA, QubitUse qbuB
  |
    // they are in the same file
    opA.getLocation().getFile() = file and
    opB.getLocation().getFile() = file and
    // they belong to the same circuit
    opA.getQuantumCircuit() = qc and
    opB.getQuantumCircuit() = qc and
    // connect bit use and the two gates
    qbuA.getAGate() = opA and
    qbuB.getAGate() = opB and
    // and they act on the same position (qubit index)
    qbuA.getAnIndex() = i and
    qbuB.getAnIndex() = i and
    // exclude undefined bit indices
    i >= 0 and
    (
      // they act on the same register
      // (if there is one explicitely instantiated)
      qbuA.getARegister() = qbuB.getARegister()
      or
      // or they act on the single quantum register of a circuit
      // experessed implicitely with e.g. QuantumCircuit(4)
      count(QuantumRegisterV2 reg | reg = qc.getAQuantumRegister() | reg) = 0 and
      qc.getNumberOfQubits() > 0
      or
      // or they act on the same quantum register but one uses the
      // integer only and the other uses the register object
      // this is unambiguous only with a single register
      // qc.h(0)
      // qc.x(qreg[0])
      count(QuantumRegisterV2 reg | reg = qc.getAQuantumRegister() | reg) = 1 and
      qc.getNumberOfQubits() > 0
    )
  )
}

/** Holds if the three operators refer to the same qubit. */
bindingset[opA, opB, opC, file, i]
pragma[inline]
predicate manipulateSameQubit(QuantumOperator opA, QuantumOperator opB, QuantumOperator opC, QuantumCircuit qc, int i, File file) {
  manipulateSameQubit(opA, opB, qc, i, file) and
  manipulateSameQubit(opB, opC, qc, i, file)
}


/** Holds if the second operator may DIRECTLY follow the first one (with no operators in between). */
bindingset[start, file]
pragma[inline]
predicate mayFollowDirectly(QuantumOperator start, QuantumOperator end, File file, int qubitIndex) {
  // mayFollow(start, end, file, qubitIndex) and
  // // check that they are immiatelly consecutive
  // not exists(QuantumOperator other |
  //   sortedInOrder(start, other, end, file, qubitIndex)
  // )
  exists(
    ControlFlowNode startNode,
    ControlFlowNode endNode,
    ControlFlowNode circNode,
    QuantumCircuit qc
  |
    startNode = start.getNode() and
    endNode = end.getNode() and
    circNode = qc.getNode() and
    // they all refer to the same qubit
    manipulateSameQubit(start, end, qc, qubitIndex, file) and
    // they are connected: start > end
    startNode.strictlyReaches(endNode) and
    // there is nothing in between
    not exists(
      QuantumOperator intermediate,
      ControlFlowNode intermediateNode
    |
      intermediateNode = intermediate.getNode() and
      // the node must be different from both at the same time
      intermediateNode != startNode and
      intermediateNode != endNode and
      // they all refer to the same qubit
      manipulateSameQubit(start, intermediate, end, qc, qubitIndex, file) and
      // such that it is between start and end
      startNode.strictlyReaches(intermediateNode) and
      intermediateNode.strictlyReaches(endNode)
    ) and
    // EXTRA PRECISION
    // they refer to the same circuit instance
    circNode.strictlyReaches(startNode) and
    circNode.strictlyReaches(endNode) and
    // we do not want a situation where the order is:
    // start >> initialization >> end
    // because they would not refer to the same circuit anymore
    not startNode.strictlyReaches(circNode)
  )
}


bindingset[start, intermediate, end, file, qubitIndex]
pragma[inline]
predicate sortedInOrder(
  QuantumOperator start, QuantumOperator intermediate, QuantumOperator end, File file,
  int qubitIndex
) {
  // they are in the same file
  start.getLocation().getFile() = file and
  intermediate.getLocation().getFile() = file and
  end.getLocation().getFile() = file and
  // they are in the same function
  // case: other >> intermediate >> this
  // exclude case: other >> this >> intermediate
  // exclude case: intermediate >> other >> this
  // exclude (loop) case: this >> other >> intermediate >> this
  // they are all in the same file
  exists(
    ControlFlowNode startNode, ControlFlowNode intermediateNode, ControlFlowNode endNode,
    QuantumCircuit qc
  |
    // the control flow and the gates are connected
    startNode = start.getNode() and
    intermediateNode = intermediate.getNode() and
    endNode = end.getNode() and
    // they work on the same qubit
    qubitIndex = start.getATargetQubit() and
    qubitIndex = intermediate.getATargetQubit() and
    qubitIndex = end.getATargetQubit() and
    // they are in the same circuit
    qc = start.getQuantumCircuit() and
    qc = intermediate.getQuantumCircuit() and
    qc = end.getQuantumCircuit() and
    // they are connected
    startNode.strictlyReaches(intermediateNode) and
    intermediateNode.strictlyReaches(endNode)
  )
}
// /** A node representing a use of a qubit in a DAG. */
// newtype TBitUseNode =
//   TBitUseAppend(QuantumOperatorViaAppend op, AppendCall appendCall, QuantumCircuit circ, OperatorSpecificationObjectName name, DataFlow::Node bitUse) {
//     op =
//       API::moduleImport("qiskit")
//           .getMember("circuit")
//           .getMember("library")
//           .getMember(name)
//           .getACall()
//     and
//     appendCall =
//       circ.getAnAttributeRead("append")
//         .getACall()
//     and
//     op = appendCall
//       .(API::CallNode)
//       .getParameter(0, "instruction")
//       .asSink()
//       // .getAValueReachingSink()
//     and
//     bitUse =
//       appendCall
//       .(API::CallNode)
//       .getParameter(1, "qargs")
//       .asSink()
//       // .getAValueReachingSink()
//   }
//   or
//   TBitUseAttribute(QuantumOperatorViaAttribute op, OperatorSpecificationAttributeName name) {
//     exists(QuantumCircuit circ |
//       // detect qc.h(0)
//       op = circ.getAnAttributeRead(name).getACall()
//     )
//   }
// /** A node representing a use of a qubit in a DAG. */
// class BitUseNode extends TBitUseNode {
//   // case: qc.append(HGate(), [0]) >> HGate
//   QuantumOperatorViaAppend opAppend;
//   OperatorSpecificationObjectName nameOpObject;
//   // case: qc.h(0) >> h
//   QuantumOperatorViaAttribute opAttribute;
//   OperatorSpecificationAttributeName nameOpAttribute;
//   string toString() {
//     this instanceof TBitUseAppend and
//     result = nameOpObject
//     or
//     this instanceof TBitUseAttribute and
//     result = nameOpAttribute
//   }
// }
