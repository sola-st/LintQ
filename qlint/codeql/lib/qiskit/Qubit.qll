
// import qiskit.Circuit
// import qiskit.Register
// import tutorial
// // QubitUsed
// // derived from a Quantum gate used and its argument
// // implicitely connected to a default QuantumRegister via an index
// // e.g. qc.h(0)
// // or explicitely connected to a QuantumRegister via quantum_register[0]
// // e.g. qc.h(quantum_register[0])
// // TODO rename QubitAccess
// class QubitUsedInteger extends IntegerLiteral {
//     QubitUsedInteger() {
//         exists(
//             Gate gate
//             |
//             this = gate.getATargetQubit()
//         )
//     }
//     Gate getGate() {
//         exists(
//             Gate gate
//             |
//             this = gate.getATargetQubit()
//             |
//             result = gate
//         )
//     }
//     int getQubitIndex() {
//         result = this.getValue()
//     }
//     QuantumRegister getQuantumRegister() {
//         // OLD
//         // exists(
//         //     QuantumCircuit circ, QuantumRegister qr
//         //     |
//         //     circ = this.getGate().getQuantumCircuit() and
//         //     qr = circ.getAQuantumRegister() and
//         //     count(QuantumRegister reg | reg = circ.getAQuantumRegister() | reg) = 1
//         //     |
//         //     result = qr
//         // )
//         // or
//         exists(
//             QuantumRegister qr, SubscriptNode subscr
//             |
//             qr.getVar().getAUse() = subscr.getObject()
//             and
//             this = subscr.getNode().getIndex()
//             |
//             result = qr
//         )
//         // NEW
//         // exists(
//         //     QuantumRegister qr,
//         //     QuantumCircuit circ,
//         //     Gate gate
//         //     |
//         //     // this qubit is used in a gate
//         //     this = gate.getATargetQubit()
//         //     // the register belongs to the circuit
//         //     and circ.getAQuantumRegister() = qr
//         //     // the gate belongs to the circuit
//         //     and circ = gate.getQuantumCircuit()
//         //     and (
//         //         // either it is the only register in the circuit
//         //         // count(circ.getAQuantumRegister()) = 1
//         //         // or
//         //         (
//         //             exists(
//         //                 SubscriptNode subscr
//         //                 |
//         //                 // or we look for the register variable which
//         //                 // is used to access the qubit (IntegerLiteral)
//         //                 qr.getVar().getAUse() = subscr.getObject()
//         //                 and this = subscr.getNode().getIndex()
//         //             )
//         //         )
//         //     )
//         //     |
//         //     result = qr
//         // )
//     }
// }
// QubitUsed
// derived from a Quantum gate used and its argument
// implicitely connected to a default QuantumRegister via an index
// e.g.  qc.h(0)
// or explicitely connected to a QuantumRegister via quantum_register[0]
// e.g. qc.h(quantum_register[0])
// .(Subscript).getIndex().(IntegerLiteral).getValue() = i
// class QubitUsed extends SubscriptNode {
//     // get a quantum register node which is used in a quantum gate
//     // qc.h(quantum_register[0])
//     QubitUsed() {
//         exists(
//             QuantumRegister qr, Variable reg_var
//             |
//             qr.getVar() = reg_var
//             and
//             reg_var.getAUse() = this.getObject()
//         ) or
//         exists(
//             QuantumRegister qr, Variable reg_var
//             |
//             qr.getVar() = reg_var
//             and
//             reg_var.getAUse() = this.getObject()
//         )
//     }
//     // get the index of the qubit used in the quantum gate
//     // qc.h(quantum_register[0])
//     int getQubitIndex() {
//         exists(
//             IntegerLiteral index
//             |
//             index = this.getNode().getIndex()
//             |
//             result = index.getValue()
//         )
//     }
//     QuantumRegister getQuantumRegister() {
//         exists(
//             QuantumRegister qr
//             |
//             qr.getVar().getAUse() = this.getObject()
//             |
//             result = qr
//         )
//     }
// }
// QubitSpace
// derived from a QuantumRegister declaration used in a QuantumCircuit
// e.g. quantum_register = QuantumRegister(2)
// Detector improvements
// find if there is any register with QubitSpace that is not used in any QubitUsed
