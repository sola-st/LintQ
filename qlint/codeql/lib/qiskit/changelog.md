2023.03.07
- fix missing gates `crx` and `cry` in the `Gate` class
- improved the `isAppliedAfter` in the `Gate` class to remove false positives when the circuit is re-initialized during the path connecting the two gates.

2023.03.07
- added the getVar() function to the register class to get the variable name of the register.
- added the QubitUsed abstraction to keep track of each access to a qubit, considering also the register it belongs to.

2023.03.03
- refactoring model: java style method calls
- refactoring of register to share part of it.
- removed GateQuantumOperation, substituted with a predicate isMeasurement() on a general gate.

2023.02.28
- refactoring of the Gate class. Remove the tracking of the for loop.

2023.02.21
- improve the efficiency of get_a_target_qubit() of the gate class (remove extra QuantumCircuit variable)
- improved the precision of get_a_target_qubit() of the gate class to account also for qubits which are operated on in a loop (e.g. qc.h(i)).


2023.02.20
- added is_subcircuit() method to Circuit class (by checking if it flows to the compose() or append() method of another circuit object)
- improved get_total_num_qubits() and get_total_num_bits() methods of Circuit class (by considering also the registers added add_register() method)
- in Circuit refactored get_num_qubits_with_integers() and get_num_bits_with_integers() methods to return the static size of the circuit (i.e. QuantumCircuit(2,2) or n = 7; m = 2, QuantumCircuit(n, m)).
- improved get_num_bits() and get_num_qubits() methods of Register class (by considering also local flows of integer variables, e.g. x = 5; QuantumRegister(x))