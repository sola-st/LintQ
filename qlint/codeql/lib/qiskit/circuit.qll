import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.register


class QuantumCircuit extends DataFlow::CallCfgNode {
    QuantumCircuit() {
        this = API::moduleImport("qiskit").getMember("QuantumCircuit").getACall()
    }

    int get_num_qubits() {
        exists(IntegerLiteral num_qubits |
            num_qubits = this.getArg(0).asExpr() and
            result = num_qubits.getValue()
        ) or
        exists( QuantumRegister qntReg|
            // there is a dataflow between the quantum register and the quantum circuit
            exists(int i |
                qntReg.flowsTo(this.getArg(i)))
            and
            // the number of qubits is the number of qubits in the quantum register
            result = qntReg.get_num_qubits()
        )
    }

    int get_num_bits() {
        exists(IntegerLiteral num_bits |
            num_bits = this.getArg(1).asExpr() and
            result = num_bits.getValue()
        ) or
        exists( ClassicalRegister clsReg|
            // there is a dataflow between the classical register and the quantum circuit
            exists(int i |
                clsReg.flowsTo(this.getArg(i)))
            and
            // the number of bits is the number of bits in the classical register
            result = clsReg.get_num_bits()
        )
    }

    int get_total_num_bits() {
        result = sum(ClassicalRegister clsReg, int i |
            clsReg.flowsTo(this.getArg(i)) |
            clsReg.get_num_bits())
    }

    int get_total_num_qubits() {
        result = sum(QuantumRegister qntReg, int i |
            qntReg.flowsTo(this.getArg(i)) |
            qntReg.get_num_qubits())
    }

}