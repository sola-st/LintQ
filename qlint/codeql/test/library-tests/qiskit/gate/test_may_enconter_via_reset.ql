import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Circuit
import qiskit.Gate
import qiskit.Qubit
import qiskit.QuantumDataFlow

from QuantumOperator opBefore, QuantumOperator opAfter, Reset resetIntermediate, int sharedQubit
where
  not opBefore.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%") and
  not opAfter.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%") and
  sharedQubit = opBefore.getATargetQubit() and
  sortedInOrder(opBefore, resetIntermediate, opAfter, opBefore.getLocation().getFile(), sharedQubit)
select opBefore, resetIntermediate, opAfter,
  "Gate: '" + opBefore.getGateName() + "' on qubit " + sharedQubit + " is followed by gate: '" +
    resetIntermediate.getGateName() + "' and then by gate: '" + opAfter.getGateName() + "'."
