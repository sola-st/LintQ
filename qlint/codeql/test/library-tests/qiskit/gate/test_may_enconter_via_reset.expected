| concept_reset_gate.py:8:1:8:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'reset' and then by gate: 'h'. |
| concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 0 is followed by gate: 'reset' and then by gate: 'h'. |
| concept_reset_gate.py:10:1:10:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset' and then by gate: 'h'. |
| concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset' and then by gate: 'x'. |
| concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset' and then by gate: 'x'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset' and then by gate: 'x'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset' and then by gate: 'x'. |
| concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset' and then by gate: 'x'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset' and then by gate: 'x'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset' and then by gate: 'cx'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset' and then by gate: 'measure'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset' and then by gate: 'reset'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset' and then by gate: 'reset'. |
